lib/os/onoff.c

```diff
@@ -0,0 +1,627 @@
+	case ONOFF_CLIENT_NOTIFY_SIGNAL:
+		if (cli->async.signal == NULL) {
+			rv = -EINVAL;
+		}
+		break;
+#endif /* CONFIG_POLL */
+	case ONOFF_CLIENT_NOTIFY_ANONYMOUS:
+		break;
+	default:
+		rv = -EINVAL;
+		break;
+	}
+
+	/* Clear the result here instead of in all callers. */
+	if (rv == 0) {
+		cli->result = 0;
+		rv = mode;
+	}
+
+	return rv;
+}
+
+int onoff_service_init(struct onoff_service *srv,
+		       onoff_service_transition_fn start,
+		       onoff_service_transition_fn stop,
+		       onoff_service_transition_fn reset,
+		       u32_t flags)
+{
+	if ((flags & SERVICE_CONFIG_FLAGS) != flags) {
+		return -EINVAL;
+	}
+
+	if ((start == NULL) || (stop == NULL)) {
+		return -EINVAL;
+	}
+
+	*srv = (struct onoff_service)ONOFF_SERVICE_INITIALIZER(start, stop, reset, flags);
+
+	return 0;
+}
+
+static void notify_one(struct onoff_service *srv,
+		       struct onoff_client *cli,
+		       int res)
+{
+	/* Store the result, and notify if requested. */
+	cli->result = res;
+	switch(cli->flags & CLIENT_NOTIFY_MASK) {
+	case ONOFF_CLIENT_NOTIFY_NO_WAIT:
+	case ONOFF_CLIENT_NOTIFY_ANONYMOUS:
+		break;
+	case ONOFF_CLIENT_NOTIFY_CALLBACK:
+		cli->async.callback.handler(srv, cli,
+					    cli->async.callback.user_data, res);
+		break;
+#if CONFIG_POLL
+	case ONOFF_CLIENT_NOTIFY_SIGNAL:
+		k_poll_signal_raise(cli->async.signal, res);
+		break;
+#endif /* CONFIG_POLL */
+	default:
+		__ASSERT_NO_MSG(false);
+	}
+}
+
+static void notify_all(struct onoff_service *srv,
+		       sys_slist_t *list,
+		       int res)
+{
+	while (!sys_slist_is_empty(list)) {
+		sys_snode_t *node = sys_slist_get_not_empty(list);
+		struct onoff_client *cli =
+			CONTAINER_OF(node,
+				     struct onoff_client,
+				     node);
+
+		notify_one(srv, cli, res);
+	}
+}
+
+static void onoff_stop_notify(struct onoff_service *srv,
+			      int res);
+
+static void onoff_start_notify(struct onoff_service *srv,
+			       int res)
+{
+	bool stop = false;
+	k_spinlock_key_t key = k_spin_lock(&srv->lock);
+	sys_slist_t clients = srv->clients;
+
+	/* Can't have a queued releaser during start */
+	__ASSERT_NO_MSG(srv->releaser == NULL);
+
+	/* If the start failed log an error and leave the rest of the
+	 * state in place for diagnostics.
+	 *
+	 * If the start succeeded record a reference for all clients
+	 * and set the state to ON.  But if there are no clients,
+	 * initiate a transition to off.
+	 *
+	 * In either case reset the client queue and notify all
+	 * clients of operation completion.
+	 */
+	if (res < 0) {
+		srv->flags &= ~SERVICE_STATE_TRANSITION;
+		srv->flags |= ONOFF_SERVICE_HAS_ERROR;
+	} else {
+		sys_snode_t *node;
+		unsigned int refs = 0U;
+
+		set_service_state(srv, SERVICE_STATE_ON);
+
+		SYS_SLIST_FOR_EACH_NODE(&clients, node) {
+			refs += 1U;
+		}
+		refs += get_anonymous_count(srv->flags);
+		srv->flags = set_anonymous_count(srv->flags, 0U);
+
+		/* Mark an error if reference count would overflow. */
+		if (srv->refs > (SERVICE_REFS_MAX - refs)) {
```

Comment:
it is quite unlikely that u16 will be reached. Maybe it would be better to just add `__ASSERT_NO_MSG(srv->refs < MAX_CLIENTS)` to `SYS_SLIST_FOR_EACH_NODE`. Number of anonymous count is controlled in request function.

Reply:
No, at this point I'm leaving open the possibility of reducing the size to `u8_t` and that could be overflowed.